File: dune/state/rounds/spice.py
from dune.state.rounds import RoundState, StageState, SubStageState


class HandleWormForces(SubStageState):
    substage = "forces"

    def __init__(self):
        self.karama_passes = []


class ResolveWormStage(StageState):
    stage = "worm"

    def __init__(self, factions):
        self.factions = factions
        self.faction_turn = None

    def visible(self, game_state, faction):
        visible = super().visible(game_state, faction)
        visible["factions"] = self.factions
        visible["faction_turn"] = self.faction_turn
        return visible


class SpiceRound(RoundState):
    round = "spice"

    def __init__(self):
        self.stage = "fremen-worm-karama"
        self.drawn_card = False
        self.needs_nexus = False

    def visible(self, game_state, faction):
        visible = super().visible(game_state, faction)
        visible["stage_state"] = self.stage_state.visible(game_state, faction) if self.stage_state else None
        return visible

File: duneclient/src/rounds/spice.js
import React from 'react';

import FactionOrder from '../components/faction-order';

export default function Spice({roundState}) {
    if (!roundState.stage_state || roundState.stage_state.stage !== "worm") {
        return [];
    }

    return <FactionOrder factions={roundState.stage_state.factions.map((faction)=>{
        return {
            faction: faction,
            active: faction === roundState.stage_state.faction_turn
        };
    })} />;
};

